staging/src/k8s.io/kubectl/pkg/cmd/rollout/rollout_pause.go

```diff
@@ -152,7 +155,14 @@ func (o *PauseOptions) RunPause() error {
 		allErrs = append(allErrs, err)
 	}
 
-	for _, patch := range set.CalculatePatches(infos, scheme.DefaultJSONEncoder(), set.PatchFn(o.Pauser)) {
+	patches := set.CalculatePatches(infos, scheme.DefaultJSONEncoder(), set.PatchFn(o.Pauser))
+
+	if len(patches) == 0 {
```

Comment:
A bit tricky, because this does solve one problem and causes another.

This is what the output looks like before this PR:

```
# When there are no StatefulSets to pause
> kubectl rollout pause statefulset

# When there is at least 1 deployment
> kubectl rollout pause deployment
deployment.apps/nginx paused

# When the given deployment doesn't exist
> kubectl rollout pause deployment/abc
Error from server (NotFound): deployments.apps "abc" not found
```

This is the output with this PR:
```
# When there are no StatefulSets to pause
> kubectl rollout pause statefulset
No resources found in default namespace.

# When there is at least 1 deployment
> kubectl rollout pause deployment
deployment.apps/nginx paused

# When the given deployment doesn't exist
> kubectl rollout pause deployment/abc
No resources found in default namespace.
Error from server (NotFound): deployments.apps "abc" not found

# When the label selector matches nothing
> kubectl rollout pause deployment -l=app=abc
No resources found in polaris namespace.
```

The problem with the current version is that attempting to pause/resume/restart all resources in a namespace by not specifying the resource name doesn't show any error when nothing matches. i.e. `kubectl rollout pause statefulset` doesn't error if there are no StatefulSets to pause.

With this PR, it will show an error because nothing changed. However, it will show 2 errors when a specific resource is given and doesn't exist.

With your suggested fix, we'd be back to problem #1 where no error shows when nothing matches (and when no label selectors were given). What I'm thinking is that maybe we should take inspiration of other commands like `kubectl delete` that requires either, `--all` or `-l` when affecting multiple resources. For example:

```
> kubectl rollout pause statefulset --all
No resources found in default namespace.

> kubectl rollout pause deployment
error: resource(s) were provided, but no name, label selector, or --all flag specified

> kubectl rollout pause deployment -l=app=abc
No resources found in default namespace.

> kubectl rollout pause deployment
deployment.apps/nginx paused
```

Seems like that would be more consistent and easier to interpret what the user is trying to do from the code. However that would be a breaking change for anyone using `kubectl rollout pause [deployment/statefulset/etc.]` without `--all` or a specific resource. Would this be better discussed in a KEP?

Reply:
@seans3 please take a look

Reply:
To add to what I said above, I think not having `--all` required was an oversight. `kubectl rollout restart deployment` can be just as destructive as `kubectl delete pod` if the `--all` flag wasn't required.